segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/OperationProcessor.java

```diff
@@ -392,66 +401,213 @@ private boolean isFatalException(Throwable ex) {
+         * by the given args has been sealed and is about to be committed. This checkpoint will mark a point in the
+         * OperationMetadataUpdater that corresponds to the end of that DataFrame.
+         *
+         * @param commitArgs The CommitArgs to create a checkpoint for.
          */
-        boolean hasPending() {
-            return !this.pendingOperations.isEmpty();
+        void checkpoint(DataFrameBuilder.CommitArgs commitArgs) {
+            synchronized (this.lock) {
+                commitArgs.setIndexKey(this.metadataUpdater.sealTransaction());
+                this.metadataTransactions.addLast(commitArgs);
+            }
         }
 
         /**
          * Commits all pending Metadata changes, assigns a TruncationMarker and acknowledges all the pending operations.
          *
          * @param commitArgs The Data Frame Commit Args that triggered this action.
-         * @throws DataCorruptionException When the operation has been committed, but failed to be accepted into the In-Memory log.
          */
-        void commit(DataFrameBuilder.DataFrameCommitArgs commitArgs) throws DataCorruptionException {
-            log.debug("{}: CommitSuccess (OperationCount = {}).", this.traceObjectId, this.pendingOperations.size());
-
-            // Record the Truncation marker and then commit any changes to metadata.
-            this.metadataUpdater.recordTruncationMarker(commitArgs.getLastStartedSequenceNumber(), commitArgs.getLogAddress());
-            this.metadataUpdater.commit();
-
-            // Acknowledge all pending entries, in the order in which they are in the queue. It is important that we ack entries in order of increasing Sequence Number.
-            while (this.pendingOperations.size() > 0 && this.pendingOperations.peek().getOperation().getSequenceNumber() <= commitArgs.getLastFullySerializedSequenceNumber()) {
-                CompletableOperation e = this.pendingOperations.poll();
-                try {
-                    this.logUpdater.process(e.getOperation());
-                } catch (Throwable ex) {
-                    log.error("{}: OperationCommitFailure ({}). {}", this.traceObjectId, e.getOperation(), ex);
-                    e.fail(ex);
-                    throw ex;
+        void commit(DataFrameBuilder.CommitArgs commitArgs) {
+            assert commitArgs.key() >= 0 : "DataFrameBuilder.CommitArgs does not have a key set";
+            log.debug("{}: CommitSuccess ({}).", this.traceObjectId, commitArgs);
+
+            List<CompletableOperation> toComplete = new ArrayList<>();
+            Map<CompletableOperation, Throwable> toFail = new HashMap<>();
+            try {
+                // Record the end of a frame in the DurableDataLog. No need for locking here, as the metadata has its own.
+                this.metadataUpdater.recordTruncationMarker(commitArgs.getLastStartedSequenceNumber(), commitArgs.getLogAddress());
```

Comment:
I changed this so it it records directly onto the metadata, which is what the metadata updater was doing anyway.